### ISSUE TITLE
0.12 locky backend: can't inherit of a class who uses joblib and is defined on another module

### ISSUE THREAD

**Issue by mhooreman** [joblib/joblib#723]
Hello,

With joblib 0.12 and the default locky backend, I can't inherit of a class who uses joblib and is defined on another module.

If I push that class in the same module, or if I go back to multiprocessing backend, it works.

It is possible, but I'm not sure, that this is a duplicate of https://github.com/joblib/joblib/issues/643. Sorry if that's the case.

The error is "``A process in the executor was terminated abruptly while the future was running or pending.``". Moreover, while showing stderr when I have the crash, I see (``TestJoblibLog`` is my child class):
``AttributeError: Can't get attribute 'TestJoblibLog' on <module 'joblib.externals.loky.backend.popen_loky_posix' from '/home/tmp_mihoo/miniconda3/lib/python3.6/site-packages/joblib/externals/loky/backend/popen_loky_posix.py'>``

So, it seems that joblib locky imports the module where joblib is called, but not the calling module which uses the results.

Please see example codes below.

Thanks!

# Example and code to reproduce

### ./in_modules.py 2>/dev/null

```
Computing chunks
Running in parallel
Hello, ['foo', 'bar', 'baz']
Extraction done
Success with default backend: ['foo', 'bar', 'baz']
Mapping foo start: 1 of 3: 33.33%
Hello, foo
Mapping foo done: 1 of 3: 33.33% in 0:00:00.180327, remaining 0:00:00.360654, end at 2018-07-25 10:52:15.640262
Mapping bar start: 2 of 3: 66.67%
Hello, bar
Mapping bar done: 2 of 3: 66.67% in 0:00:00.181148, remaining 0:00:00.090574, end at 2018-07-25 10:52:15.371003
Mapping baz start: 3 of 3: 100.00%
Hello, baz
Mapping baz done: 3 of 3: 100.00% in 0:00:00.181711, remaining 0:00:00, end at 2018-07-25 10:52:15.280992
```

### ./imported.py 2>/dev/null

```
Computing chunks
Running in parallel
Exception with default backend: A process in the executor was terminated abruptly while the future was running or pending.
Computing chunks
Running in parallel
Using multiprocessing joblib backend instead of the default one
Mapping foo start: 1 of 3: 33.33%
Hello, foo
Mapping foo done: 1 of 3: 33.33% in 0:00:00.013259, remaining 0:00:00.026518, end at 2018-07-25 10:56:25.080652
Mapping bar start: 2 of 3: 66.67%
Hello, bar
Mapping bar done: 2 of 3: 66.67% in 0:00:00.013529, remaining 0:00:00.006764, end at 2018-07-25 10:56:25.061168
Mapping baz start: 3 of 3: 100.00%
Hello, baz
Mapping baz done: 3 of 3: 100.00% in 0:00:00.013826, remaining 0:00:00, end at 2018-07-25 10:56:25.054701
Hello, ['foo', 'bar', 'baz']
Extraction done
Success with mutiprocessing backend: ['foo', 'bar', 'baz']
```

### imported.py

```python
#!/usr/bin/env python3

from wrapper import MapReduceWrapper


class TestJoblibLog(MapReduceWrapper):
    def __init__(self, *args, **kwargs):
        MapReduceWrapper.__init__(self, *args, **kwargs)

    def computeChunks(self):
        print("Computing chunks")
        return ['foo', 'bar', 'baz']

    def map(self, chunk):
        print(f"Hello, {chunk}")
        return chunk

    def reduce(self, mapResults):
        print(f"Hello, {str(mapResults)}")
        return mapResults


if __name__ == "__main__":
    try:
        print(
            "Success with default backend:",
            TestJoblibLog(backend=None).get()
        )
    except Exception as e:
        print(f"Exception with default backend: {e}")
        try:
            print(
                "Success with mutiprocessing backend:",
                TestJoblibLog(backend='multiprocessing').get()
            )
        except Exception as e:
            print(f"Exception with multiprocessing backend: {e}")
            print("No result")
```

### wrapper.py

```python
#!/usr/bin/env python3

import abc
import datetime
import joblib as jl


class MapReduceWrapper(metaclass=abc.ABCMeta):
    def __init__(self, nJobs=-1, verbose=2, backend='multiprocessing'):
        """Constructor
        - nJobs: if 1 or 0, don't use parallel processing; else, use joblib
                 parallel processing, and forward this value to the n_jobs
                 argument of the joblib.Parallel's constructor
        - verbose: 0: don't show anything, 1: show chunk name as info message,
                   2: show chunk name and progres as info message
        - backend: the joblib backend. Useful because, since jl 0.12, the
                   locky backend is used by default, and it makes some issues;
                   in this case, multiprocessing backend is better. If
                   the provided value is None, it let joblib choose the
                   backend.
        Please note that, because of parallel processing, the estimated
        progress and elapsed time are not accurate
        """
        self._nJobs = nJobs
        self._verbose = verbose
        self._useParallel = (nJobs not in (0, 1))
        self._mapStart = None
        self._backend = backend
        self._chunks = self.computeChunks()

    @abc.abstractmethod
    def computeChunks(self):
        """Returns the list of the chunks
        Will typically give a list based on child's class constructor's
        arguments
        To be implemented in the child classes
        """

    @abc.abstractmethod
    def map(self, chunk):
        """Gets data from the provided chunk and returns the result
        - chunk: the chunk identifier
        To be implemented in the child classes
        """

    @abc.abstractmethod
    def reduce(self, mapResults):
        """Reduce the data retrieven from every map call and returns the result
        - mapResults: list of map results
        To be implemented in the child classes
        """

    def _loggedMap(self, i, n, x):
        """Wrapper for map, logging the ongoing chunk + progress estimation
        - i: id of the current chunk, from 1 to the number of chunks
        - n: number of chunks
        - x: chunk identifier
        """
        if self._verbose == 2:
            print("Mapping %s start: %d of %d: %.2f%%" % (
                x,
                i,
                n,
                100 * i / n,
            ))
        elif self._verbose == 1:
            print("Mapping %s" % x)
        ret = self.map(x)

        if self._verbose == 2:
            progress = 100 * i / n
            now = datetime.datetime.now()
            elapsed = now - self._mapStart
            remaining = datetime.timedelta(
                seconds=(
                    elapsed.total_seconds() * (100 - progress) / progress
                )
            )
            endAt = now + remaining
            print(
                (
                    "Mapping %s done: %d of %d: %.2f%% in %s, remaining %s, "
                    "end at %s"
                ) % (
                    x,
                    i,
                    n,
                    100 * i / n,
                    str(elapsed),
                    str(remaining),
                    str(endAt)
                )
            )
        return ret

    def get(self):
        """Computes and returns the selected data"""
        n = len(self._chunks)
        self._mapStart = datetime.datetime.now()
        if self._useParallel:
            print("Running in parallel")
            if self._nJobs != -1:
                self.logWarning(
                    "Running %d jobs in parallel, which might be different "
                    "than one per CPU"
                )
            jlKwargs = dict(n_jobs=self._nJobs, verbose=0)
            if self._backend is not None:
                print(
                    f"Using {self._backend} joblib backend "
                    "instead of the default one"
                )
                jlKwargs['backend'] = self._backend
            xs = jl.Parallel(**jlKwargs)(
                jl.delayed(self._loggedMap)(
                    i + 1, n, chunk
                ) for i, chunk in list(enumerate(self._chunks))
            )
        else:
            self.logWarning(
                "Parallel processing disabled; running in sequence"
            )
            xs = [
                self._loggedMap(
                    i + 1, n, chunk
                ) for (
                    i, chunk
                ) in enumerate(
                    self._chunks
                )
            ]
        self._mapStart = None
        ret = self.reduce(xs)
        try:
            ret = ret.copy()
        except AttributeError:
            self.logWarning(
                f"Deep copy unavailable for return type {type(ret)}, "
                "returning direct value instead."
            )
        print("Extraction done")
        return ret
```

### in_module.py

```python
#!/usr/bin/env python3

import abc
import datetime
import joblib as jl

from vitriol.base import VitriolBase
from vitriol.algorithms.parallel import MapReduceWrapper

class MapReduceWrapper(metaclass=abc.ABCMeta):
    def __init__(self, nJobs=-1, verbose=2, backend='multiprocessing'):
        """Constructor
        - nJobs: if 1 or 0, don't use parallel processing; else, use joblib
                 parallel processing, and forward this value to the n_jobs
                 argument of the joblib.Parallel's constructor
        - verbose: 0: don't show anything, 1: show chunk name as info message,
                   2: show chunk name and progres as info message
        - backend: the joblib backend. Useful because, since jl 0.12, the
                   locky backend is used by default, and it makes some issues;
                   in this case, multiprocessing backend is better. If
                   the provided value is None, it let joblib choose the
                   backend.
        Please note that, because of parallel processing, the estimated
        progress and elapsed time are not accurate
        """
        self._nJobs = nJobs
        self._verbose = verbose
        self._useParallel = (nJobs not in (0, 1))
        self._mapStart = None
        self._backend = backend
        self._chunks = self.computeChunks()

    @abc.abstractmethod
    def computeChunks(self):
        """Returns the list of the chunks
        Will typically give a list based on child's class constructor's
        arguments
        To be implemented in the child classes
        """

    @abc.abstractmethod
    def map(self, chunk):
        """Gets data from the provided chunk and returns the result
        - chunk: the chunk identifier
        To be implemented in the child classes
        """

    @abc.abstractmethod
    def reduce(self, mapResults):
        """Reduce the data retrieven from every map call and returns the result
        - mapResults: list of map results
        To be implemented in the child classes
        """

    def _loggedMap(self, i, n, x):
        """Wrapper for map, logging the ongoing chunk + progress estimation
        - i: id of the current chunk, from 1 to the number of chunks
        - n: number of chunks
        - x: chunk identifier
        """
        if self._verbose == 2:
            print("Mapping %s start: %d of %d: %.2f%%" % (
                x,
                i,
                n,
                100 * i / n,
            ))
        elif self._verbose == 1:
            print("Mapping %s" % x)
        ret = self.map(x)

        if self._verbose == 2:
            progress = 100 * i / n
            now = datetime.datetime.now()
            elapsed = now - self._mapStart
            remaining = datetime.timedelta(
                seconds=(
                    elapsed.total_seconds() * (100 - progress) / progress
                )
            )
            endAt = now + remaining
            print(
                (
                    "Mapping %s done: %d of %d: %.2f%% in %s, remaining %s, "
                    "end at %s"
                ) % (
                    x,
                    i,
                    n,
                    100 * i / n,
                    str(elapsed),
                    str(remaining),
                    str(endAt)
                )
            )
        return ret

    def get(self):
        """Computes and returns the selected data"""
        n = len(self._chunks)
        self._mapStart = datetime.datetime.now()
        if self._useParallel:
            print("Running in parallel")
            if self._nJobs != -1:
                self.logWarning(
                    "Running %d jobs in parallel, which might be different "
                    "than one per CPU"
                )
            jlKwargs = dict(n_jobs=self._nJobs, verbose=0)
            if self._backend is not None:
                print(
                    f"Using {self._backend} joblib backend "
                    "instead of the default one"
                )
                jlKwargs['backend'] = self._backend
            xs = jl.Parallel(**jlKwargs)(
                jl.delayed(self._loggedMap)(
                    i + 1, n, chunk
                ) for i, chunk in list(enumerate(self._chunks))
            )
        else:
            self.logWarning(
                "Parallel processing disabled; running in sequence"
            )
            xs = [
                self._loggedMap(
                    i + 1, n, chunk
                ) for (
                    i, chunk
                ) in enumerate(
                    self._chunks
                )
            ]
        self._mapStart = None
        ret = self.reduce(xs)
        try:
            ret = ret.copy()
        except AttributeError:
            self.logWarning(
                f"Deep copy unavailable for return type {type(ret)}, "
                "returning direct value instead."
            )
        print("Extraction done")
        return ret


class TestJoblibLog(MapReduceWrapper):
    def __init__(self, *args, **kwargs):
        MapReduceWrapper.__init__(self, *args, **kwargs)

    def computeChunks(self):
        print("Computing chunks")
        return ['foo', 'bar', 'baz']

    def map(self, chunk):
        print(f"Hello, {chunk}")
        return chunk

    def reduce(self, mapResults):
        print(f"Hello, {str(mapResults)}")
        return mapResults


if __name__ == "__main__":
    try:
        print(
            "Success with default backend:",
            TestJoblibLog(backend=None).get()
        )
    except Exception as e:
        print(f"Exception with default backend: {e}")
        try:
            print(
                "Success with mutiprocessing backend:",
                TestJoblibLog(backend='multiprocessing').get()
            )
        except Exception as e:
            print(f"Exception with multiprocessing backend: {e}")
            print("No result")
```

**Comment by ogrisel** [comment]
Thanks for the report. Could you please try to make it minimal (that is as small as possible while still being able to reproduce the original problem)? That would help us a lot.

**Comment by mhooreman** [comment]
Hi @ogrisel 

Indeed, my example was over-complex. Here's a better one:

- module `wrapper` defininig a `AbstractWrapper` class. This class:
    - Does "parallel" processing of chunks
    - For each chunk, calls `self._mapSugar`, which calls the abstract `self.map`. That sugar, in addition to the value to map, gives also some progress related information (like item x on y, etc. ... in my own logging system (not the joblib verbose))
    - The constructor can take a `nJobs` and `backend` arguments (`nJobs` is your `n_jobs`). If they are `None`, they are not sent to your constructor
- `test.py` implements `AbstractWrapper`, and test with different combinations of `nJobs` and `wrapper`

**Remark:** As you will see in code comments below, there is no issue if we call directly `self.map` without passing through `self._mapSugar`

Here's the standard output of `./test.py` (`2>&1` for clarity, but the details might be interesting):
```
Testing with nJobs=None and backend=None
Success
Testing with nJobs=None and backend=multiprocessing
Success
Testing with nJobs=-1 and backend=None
Failure: A process in the executor was terminated abruptly while the future was running or pending.
Testing with nJobs=-1 and backend=multiprocessing
Success
```

And the code:

```python
# ############################################################################
# # This is main.py which MUST be in a separate file                         #
# ############################################################################
#!/usr/bin/env python3

from wrapper import AbstractWrapper

import time

class ConcreteWrapper(AbstractWrapper):
    def map(self, x):
        time.sleep(0.1)
        return x

def test(concreteWrapper):
    def _testCombination(nJobs, backend):
        print(f"Testing with nJobs={nJobs} and backend={backend}")
        kw = {}
        if backend is not None:
            kw['backend'] = backend
        if nJobs is not None:
            kw['n_jobs'] = nJobs
        try:
            concreteWrapper(**kw).get()
            print("Success")
            return True
        except Exception as e:
            print(f"Failure: {e}")
            return False
    for aa in (
        (None, None),
        (None, 'multiprocessing'),
        (-1, None),
        (-1, 'multiprocessing'),
    ):
        _testCombination(*aa)
```

```python
# ############################################################################
# # This is wrapper.py which MUST be in a separate file                      #
# ############################################################################
import abc
import joblib as jl

class AbstractWrapper(metaclass=abc.ABCMeta):
    def __init__(self, **jlKwargs):
        self._jlKwargs = jlKwargs

    def get(self):
        xx = range(10)
        # No issue if we use the code below
        # return(jl.Parallel(**self._jlKwargs)(jl.delayed(self.map)(x) for x in xx))
        n = len(xx)
        jl.Parallel(**self._jlKwargs)(
            jl.delayed(self._mapSugar)(
                i + 1, n, chunk
            ) for i, chunk in list(enumerate(xx))
        )

    def _mapSugar(self, i, n, x):
        return self.map(x)

    @abc.abstractmethod
    def map(self, x):
        pass
```

**Comment by ogrisel** [comment]
Thank you very much @mhooreman for your effort in making this minimal, this is really helpful. I will try to find the time to investigate soon.

**Comment by ogrisel** [comment]
Actually, I think I understand: the fact that `_mapSugar` is defined in an importable module makes joblib think that it can use the regular (fast) python pickle, while in fact it's a method from an instance of a class defined interactively in the `__main__` module which requires to use the cloudpickle wrapper.

Our introspection logic can probably be improved to deal with this case.

**Comment by ogrisel** [comment]
I have pushed a fix in #726 to fix the detection of bound methods of an instance of an interactively defined class even when they inherit from an importable class.

However, your particular test still fails because cloudpickle can't pickle _abc_data objects.

**Comment by mhooreman** [comment]
Thanks for the hard work.

Indeed, this construction is not straightforward at all but ... it's needed
in my implementation of a easy to use and transparent map/reduce framework.

Thanks again

Le jeu. 26 juil. 2018 18:34, Olivier Grisel <notifications@github.com> a
écrit :

**Comment by ogrisel** [comment]
I see no reason why _abc_data would not be picklable. I think this need a fix in cloudpickle.

**Comment by mhooreman** [comment]
Thanks @ogrisel.

What does that means concretely? I really would like to help, but I'm not confident to write a cloudpickle ticket for that because I don't know enough the mechanisms that you use in joblib.

**Comment by ogrisel** [comment]
Actually, there is no problem on Python 3.6, just on Python 3.7. And apparently this bug has already been reported as: https://github.com/cloudpipe/cloudpickle/issues/180

**Comment by ogrisel** [comment]
For the record here is the traceback:

```python-traceback
Traceback (most recent call last):
  File "test_subclass.py", line 36, in <module>
    test(ConcreteWrapper)
  File "test_subclass.py", line 33, in test
    _testCombination(*aa)
  File "test_subclass.py", line 21, in _testCombination
    concreteWrapper(**kw).get()
  File "/home/ogrisel/tmp/wrapper.py", line 17, in get
    ) for i, chunk in list(enumerate(xx))
  File "/home/ogrisel/code/joblib/joblib/parallel.py", line 994, in __call__
    self.retrieve()
  File "/home/ogrisel/code/joblib/joblib/parallel.py", line 897, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/opt/python3.7/lib/python3.7/multiprocessing/pool.py", line 657, in get
    raise self._value
  File "/opt/python3.7/lib/python3.7/multiprocessing/pool.py", line 431, in _handle_tasks
    put(task)
  File "/home/ogrisel/code/joblib/joblib/pool.py", line 158, in send
    CustomizablePickler(buffer, self._reducers).dump(obj)
  File "/home/ogrisel/code/joblib/joblib/parallel.py", line 290, in __getstate__
    for func, args, kwargs in self.items]
  File "/home/ogrisel/code/joblib/joblib/parallel.py", line 290, in <listcomp>
    for func, args, kwargs in self.items]
  File "/home/ogrisel/code/joblib/joblib/parallel.py", line 278, in _wrap_non_picklable_objects
    wrapped_obj = CloudpickledObjectWrapper(obj)
  File "/home/ogrisel/code/joblib/joblib/parallel.py", line 208, in __init__
    self.pickled_obj = dumps(obj)
  File "/home/ogrisel/code/joblib/joblib/externals/cloudpickle/cloudpickle.py", line 881, in dumps
    cp.dump(obj)
  File "/home/ogrisel/code/joblib/joblib/externals/cloudpickle/cloudpickle.py", line 268, in dump
    return Pickler.dump(self, obj)
  File "/opt/python3.7/lib/python3.7/pickle.py", line 437, in dump
    self.save(obj)
  File "/opt/python3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/home/ogrisel/code/joblib/joblib/externals/cloudpickle/cloudpickle.py", line 661, in save_instancemethod
    self.save_reduce(types.MethodType, (obj.__func__, obj.__self__), obj=obj)
  File "/opt/python3.7/lib/python3.7/pickle.py", line 638, in save_reduce
    save(args)
  File "/opt/python3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/python3.7/lib/python3.7/pickle.py", line 771, in save_tuple
    save(element)
  File "/opt/python3.7/lib/python3.7/pickle.py", line 549, in save
    self.save_reduce(obj=obj, *rv)
  File "/opt/python3.7/lib/python3.7/pickle.py", line 633, in save_reduce
    save(cls)
  File "/opt/python3.7/lib/python3.7/pickle.py", line 518, in save
    self.save_global(obj)
  File "/home/ogrisel/code/joblib/joblib/externals/cloudpickle/cloudpickle.py", line 636, in save_global
    return self.save_dynamic_class(obj)
  File "/home/ogrisel/code/joblib/joblib/externals/cloudpickle/cloudpickle.py", line 495, in save_dynamic_class
    save(clsdict)
  File "/opt/python3.7/lib/python3.7/pickle.py", line 504, in save
    f(self, obj) # Call unbound method with explicit self
  File "/opt/python3.7/lib/python3.7/pickle.py", line 856, in save_dict
    self._batch_setitems(obj.items())
  File "/opt/python3.7/lib/python3.7/pickle.py", line 882, in _batch_setitems
    save(v)
  File "/opt/python3.7/lib/python3.7/pickle.py", line 524, in save
    rv = reduce(self.proto)
TypeError: can't pickle _abc_data objects
```

**Comment by ogrisel** [comment]
@mhooreman have you tried with Python 3.6 and joblib 0.12.2 ? Because I could not reproduce it when adapting your code to make it run on my machine (you did not include anything that calls the `test` function in the main script.

**Comment by ogrisel** [comment]
You need to check with joblib 0.12.2 that includes the fix from #726. joblib 0.12.1 does not have this fix.

**Comment by mhooreman** [comment]
@ogrisel 

I've checked with joblib 0.12.2 on a dedicated conda env and manuall installation of joblib.

I've added (the forgotten) following to test.py: 
```
if __name__ == "__main__":
    test(ConcreteWrapper)
```

And... it doesn't works .. and it is even worst because **multiprocessing is now making issues**:

```
$ ./test.py 
Testing with nJobs=None and backend=None
Success
Testing with nJobs=None and backend=multiprocessing
Success
Testing with nJobs=-1 and backend=None
Failure: Could not pickle the task to send it to the workers.
Testing with nJobs=-1 and backend=multiprocessing
Failure: can't pickle _abc_data objects
```

**Comment by ogrisel** [comment]
Thanks for the confirmation.

**Comment by mhooreman** [comment]
You're welcome, Olivier, but I'm a bit lost now.

So, if my understanding is correct, neitjer the new, nor the old backend
works with 3.7.

Do you know the plans to fix it?

Sorry for the stupid questions.

Michaël

Le dim. 5 août 2018 20:44, Olivier Grisel <notifications@github.com> a
écrit :

**Comment by ogrisel** [comment]
Fixed by #756.
